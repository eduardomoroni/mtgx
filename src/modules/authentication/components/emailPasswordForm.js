// @flow

import React, { Component } from 'react'
import { PropTypes } from 'prop-types'
import I18n from 'react-native-i18n'
import { View } from 'react-native'
import { Input, Button } from 'nachos-ui'
import { styles } from './styles/emailPasswordForm.style'

export class EmailPasswordForm extends Component {
  state = {
    email: '',
    password: ''
  }

  static propTypes = {
    buttonText: PropTypes.string.isRequired,
    buttonCallback: PropTypes.func.isRequired
  }

  render () {
    const { buttonCallback, buttonText } = this.props
    const { email, password } = this.state

    return (
      <View>
        <Input
          icon='ios-mail-outline'
          style={[styles.input]}
          placeholder={I18n.t('EMAIL')}
          value={this.state.email}
          onChangeText={email => this.setState({email})}
        />
        <Input
          icon='ios-lock-outline'
          style={[styles.input]}
          placeholder={I18n.t('PASSWORD')}
          value={this.state.password}
          onChangeText={password => this.setState({password})}
        />
        <Button
          kind='squared'
          type='success'
          style={[styles.button]}
          onPress={() => buttonCallback(email, password)}
        >
          {buttonText}
        </Button>
      </View>
    )
  }
}
