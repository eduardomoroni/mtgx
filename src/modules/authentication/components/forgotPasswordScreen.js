// @flow

import React, { Component } from 'react'
import I18n from 'react-native-i18n'
import PropTypes from 'prop-types'
import { View } from 'react-native'
import { H1, Button, Input } from 'nachos-ui'
import { styles } from './styles/forgotPassword.style'

export class ForgotPasswordScreen extends Component {
  state = {
    email: ''
  }

  static propTypes = {
    resetPassword: PropTypes.func.isRequired
  }

  onForgotPasswordPressed = async () => {
    this.props.resetPassword(this.state.email)
  }

  render () {
    return (
      <View style={styles.container}>
        <H1>{I18n.t('RESET_PASSWORD')}</H1>
        <Input
          icon='ios-mail-outline'
          style={[styles.input]}
          placeholder={I18n.t('EMAIL')}
          value={this.state.email}
          onChangeText={email => this.setState({email})}
        />
        <Button
          kind='squared'
          type='success'
          style={[styles.button]}
          onPress={this.onForgotPasswordPressed}
        >
          {I18n.t('RESET_PASSWORD')}
        </Button>
      </View>
    )
  }
}
