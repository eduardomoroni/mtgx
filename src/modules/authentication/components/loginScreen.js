// @flow

import React, { Component } from 'react'
import Config from 'react-native-config'
import I18n from 'react-native-i18n'
import { StyleSheet, Text, View } from 'react-native'
import { H4, Button } from 'nachos-ui'

export class LoginScreen extends Component {
  renderButton () {
    const btnStyle = { margin: 15 }
    return (
      <View>
        <H4>Example:</H4>
        <Button kind='squared' type='success' style={btnStyle}>
          Success
        </Button>
        <Button kind='squared' type='danger' style={btnStyle}>
          Danger
        </Button>
        <Button
          kind='squared'
          iconName='md-cloud-download'
          style={btnStyle}
        >
          Primary
        </Button>
        <H4 align='center'>Disabled style</H4>
      </View>
    )
  }

  render () {
    return (
      <View style={styles.container}>
        <Text style={styles.welcome}>
          Welcome to React Native!
        </Text>
        <Text style={styles.instructions}>
          To get started, edit index.ios.js
        </Text>
        <Text style={styles.instructions}>
          Press Cmd+R to reload,{'\n'}
          Cmd+D or shake for dev menu
        </Text>
        <Text style={styles.instructions}>
          {Config.TEST_ENV}
        </Text>
        <Text style={styles.instructions}>
          {I18n.t('test')}
        </Text>
        {this.renderButton()}
      </View>
    )
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
    backgroundColor: '#F5FCFF'
  },
  welcome: {
    fontSize: 20,
    textAlign: 'center',
    margin: 10
  },
  instructions: {
    textAlign: 'center',
    color: '#333333',
    marginBottom: 5
  }
})
