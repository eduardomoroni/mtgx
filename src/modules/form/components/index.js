// @flow

export * from './inputLabel'
export * from './inputPicker'
export * from './textInputForm'
export * from './switchInputForm'
export * from './modalToggle'
export * from './multiSelect'
// $FlowFixMe
export * from './numericInputForm'
// $FlowFixMe
export * from './dropdownInputForm'
