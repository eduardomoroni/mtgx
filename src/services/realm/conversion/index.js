// @flow

import _ from 'lodash'

export { toRealmCard } from './json'

const mapFormToRealm = {
  cardName: 'name CONTAINS[c]',
  cardSubType: 'subtypes.subType =',
  cardType: 'types.type =',
  cardText: 'text CONTAINS[c]',
  cardColors: 'colors.color =',
  cardSet: 'printings.printing =',
  cardArtist: 'artist CONTAINS[c]',
  cardFlavorText: 'flavor CONTAINS[c]',
  cardCollectionNumber: 'number =',
  cardRarity: 'rarity =',
  cardColorsIdentity: 'colorIdentity.colorIdentity =',
  cardToughness: 'toughness',
  cardPower: 'power',
  cardCMC: 'cmc',
  cardFormat: 'legalities.format ='
}

function arrayToQuery (array, selector, operator) {
  return array.reduce((acc, current) => {
    return acc ? `${acc} ${operator} ${selector} "${current}"` : `${selector} "${current}"`
  }, '')
}

export const inheritanceToArray = (realmRepresentation: Object) => {
  const detectKeyValue = (obj) => {
    const objKeys = _.keys(obj)
    return obj[objKeys[0]]
  }

  const withoutFieldName = _.mapValues(realmRepresentation, detectKeyValue)
  return _.toArray(withoutFieldName)
}

export const convertCardFormToRealmQueries = (cardForm) => {
  return _.mapValues(cardForm, (value, key) => {
    if (value === undefined || mapFormToRealm[key] === undefined) {
      return undefined
    } else if (Array.isArray(value)) {
      return arrayToQuery(value, mapFormToRealm[key], 'OR')
    } else if (typeof value === 'object') {
      return `${mapFormToRealm[key]} ${value.operator} ${value.number}`
    } else {
      return `${mapFormToRealm[key]} "${value}"`
    }
  })
}
