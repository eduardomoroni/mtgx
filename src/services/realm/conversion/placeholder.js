// @flow

export const placeholdersToSymbols = (cardText) => {
  const curlyBracesRegex = /[{}]+/
  return (cardText || '').split(curlyBracesRegex)
    .map(a => convertionMap[a] || a)
    .join('')
}

function unicodeChar (value) {
  return String.fromCharCode(parseInt(value, 16))
}

// https://andrewgioia.github.io/Mana/cheatsheet.html
const convertionMap = {
  'W': 'e600',
  'U': 'e601',
  'B': 'e602',
  'R': 'e603',
  'G': 'e604',
  'C': 'e904',
  '0': 'e605',
  '1': 'e606',
  '2': 'e607',
  '3': 'e608',
  '4': 'e609',
  '5': 'e60a',
  '6': 'e60b',
  '7': 'e60c',
  '8': 'e60d',
  '9': 'e60e',
  '10': 'e60F',
  '11': 'e610',
  '12': 'e611',
  '13': 'e612',
  '14': 'e613',
  '15': 'e614',
  '16': 'e62a',
  '17': 'e62b',
  '18': 'e62c',
  '19': 'e62d',
  '20': 'e62e',
  'X': 'e615',
  'Y': 'e616',
  'Z': 'e617',
  'P': 'e618',
  'S': 'e619',
  'E': 'e907'
}

// Converting convertionMap to encoded char
Object.keys(convertionMap).forEach(key => {
  convertionMap[key] = unicodeChar(convertionMap[key])
})
