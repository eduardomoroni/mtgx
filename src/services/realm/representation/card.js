// @flow

import PropTypes from 'prop-types'

const legalitiesType = PropTypes.shape({
  format: PropTypes.string.isRequired,
  legality: PropTypes.string.isRequired
})

const superTypesType = PropTypes.shape({
  superType: PropTypes.string.isRequired
})

export const rulingsType = PropTypes.objectOf(
  PropTypes.shape({
    date: PropTypes.string.isRequired,
    text: PropTypes.string.isRequired
  })
)

const foreignNameType = PropTypes.shape({
  language: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  multiverseid: PropTypes.number.isRequired
})

export const printingsType = PropTypes.shape({
  printing: PropTypes.string.isRequired
})

const colorIdentityType = PropTypes.shape({
  colorIdentity: PropTypes.string.isRequired
})

const colorsType = PropTypes.shape({
  color: PropTypes.string.isRequired
})

const subtypesType = PropTypes.shape({
  subType: PropTypes.string.isRequired
})

const typesType = PropTypes.shape({
  type: PropTypes.string.isRequired
})

export const cardType = PropTypes.shape({
  multiverseid: PropTypes.number.isRequired,
  artist: PropTypes.string.isRequired,
  id: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  number: PropTypes.string.isRequired,
  rarity: PropTypes.string.isRequired,
  type: PropTypes.string.isRequired,
  cmc: PropTypes.number,
  power: PropTypes.number,
  toughness: PropTypes.number,
  imageName: PropTypes.string,
  layout: PropTypes.string,
  manaCost: PropTypes.string,
  originalText: PropTypes.string,
  originalType: PropTypes.string,
  flavor: PropTypes.string,
  text: PropTypes.string,
  types: PropTypes.objectOf(typesType),
  subtypes: PropTypes.objectOf(subtypesType),
  colors: PropTypes.objectOf(colorsType),
  colorIdentity: PropTypes.objectOf(colorIdentityType),
  printings: PropTypes.objectOf(printingsType),
  foreignNames: PropTypes.objectOf(foreignNameType),
  rulings: rulingsType,
  superTypes: PropTypes.objectOf(superTypesType),
  legalities: PropTypes.objectOf(legalitiesType)
})
