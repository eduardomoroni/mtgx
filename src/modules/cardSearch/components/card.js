// @flow

import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { View, Text } from 'react-native'

import { placeholdersToSymbols } from '../../../services/realm/conversion/placeholder' // TODO: REVISIT
import { cardType, printingsType } from '../types/cardType'
import { styles } from './styles/card.styles'

// TODO: REVISIT
export const getLastPrinting = (printings: printingsType) => {
  const keys = Object.keys(printings)
  const lastPrinting = printings[keys[keys.length - 1]]
  return lastPrinting.printing
}

export class Card extends Component {
  static propTypes = {
    showCardText: PropTypes.bool,
    card: cardType
  }

  renderCardNameAndMana = () => {
    const { name, manaCost } = this.props.card

    return (
      <View style={styles.lineContainer}>
        <View style={styles.leftWord}>
          <Text key='cardName' style={styles.cardNameText}>
            {name}
          </Text>
        </View>
        <View style={styles.rightWord}>
          <Text key='cardManaCost' style={styles.mana}>
            {placeholdersToSymbols(manaCost)}
          </Text>
        </View>
      </View>
    )
  }

  renderCardTypeAndEdition = () => {
    const { type, printings } = this.props.card

    return (
      <View style={styles.lineContainer}>
        <View style={styles.leftWord}>
          <Text key='cardType' style={styles.cardTypeText}>
            {type}
          </Text>
        </View>
        <View style={styles.rightWord}>
          <Text key='cardEdition'>
            {getLastPrinting(printings)}
          </Text>
        </View>
      </View>
    )
  }

  renderTextAndPower = () => {
    const { card, showCardText } = this.props
    const { text, power, toughness } = card

    return (
      <View style={styles.cardTextContainer}>
        <View style={styles.cardText}>
          <Text key='cardText' style={styles.text}>
            {showCardText ? text : ''}
          </Text>
        </View>
        <View style={styles.cardPower}>
          <Text key='cardPowerAndToughness' style={styles.cardPowerToughness}>
            {power || toughness ? `${power}/${toughness}` : ''}
          </Text>
        </View>
      </View>
    )
  }

  render () {
    return (
      <View style={styles.container}>
        {this.renderCardNameAndMana()}
        {this.renderCardTypeAndEdition()}
        {this.renderTextAndPower()}
      </View>
    )
  }
}
