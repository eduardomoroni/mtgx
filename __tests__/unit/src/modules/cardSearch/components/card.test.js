import React from 'react'
import { shallow } from 'enzyme'
import { Card, getLastPrinting } from '../../../../../../src/modules/cardSearch/components/card'
import * as Conversion from '../../../../../../src/services/realm/conversion/placeholder'

const props = {
  showCardText: true,
  card: {
    name: 'CardName',
    text: 'CardText',
    manaCost: '{1}{W}{B}',
    type: 'CardType',
    power: 2,
    artist: 'EduardoMoroni',
    id: 'id',
    number: 123,
    rarity: 'Common',
    multiverseid: 12903,
    toughness: 4,
    printings: { '0': { printing: 'AER' }, '1': { printing: 'AST' } }
  }
}

const hasKey = key => {
  return n => n.key() === key
}

const findRenderedTextFromKey = (wrapper, key) => {
  return wrapper.findWhere(hasKey(key)).childAt(0).text()
}

describe('<Card />', () => {
  const {
    name,
    manaCost,
    type,
    text,
    power,
    toughness,
    printings
  } = props.card
  Conversion.placeholdersToSymbols = jest.fn(text => `placeholdersToSymbols(${text})`)
  const wrapper = shallow(<Card {...props} />)

  it('should render a snapshot', () => {
    expect(wrapper).toMatchSnapshot()
  })

  it('should render Card component', () => {
    expect(findRenderedTextFromKey(wrapper, 'cardName')).toEqual(name)
    expect(findRenderedTextFromKey(wrapper, 'cardManaCost')).toEqual(Conversion.placeholdersToSymbols(manaCost))
    expect(findRenderedTextFromKey(wrapper, 'cardType')).toEqual(type)
    expect(findRenderedTextFromKey(wrapper, 'cardEdition')).toEqual(getLastPrinting(printings))
    expect(findRenderedTextFromKey(wrapper, 'cardText')).toEqual(text)
    expect(findRenderedTextFromKey(wrapper, 'cardPowerAndToughness')).toEqual(`${power}/${toughness}`)
    expect(Conversion.placeholdersToSymbols).toHaveBeenCalledWith(props.card.manaCost)
  })

  it('should hide card text', () => {
    const wrapper = shallow(<Card {...props} showCardText={false} />)
    expect(findRenderedTextFromKey(wrapper, 'cardText')).toEqual('')

    wrapper.setProps({...props, showCardText: true})
    expect(findRenderedTextFromKey(wrapper, 'cardText')).toEqual(props.card.text)
  })

  it('should show only the lastest printing as cardEdition', () => {
    const printExamples = { '0': { printing: 'STH' }, '1': { printing: 'TPR' }, '2': { printing: 'EXP' } }
    const wrapper = shallow(<Card card={{...props.card, printings: printExamples}} />)
    expect(findRenderedTextFromKey(wrapper, 'cardEdition')).toEqual('EXP')
  })
  it('should extract last printing from realm representation', () => {
    const realmRepresentation = { '0': { printing: 'STH' }, '1': { printing: 'TPR' }, '2': { printing: 'AER' } }
    expect(getLastPrinting(realmRepresentation)).toEqual('AER')
  })
})
