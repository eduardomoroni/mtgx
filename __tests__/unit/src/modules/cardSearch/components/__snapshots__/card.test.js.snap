// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Card /> should render a snapshot 1`] = `
<View
  style={
    Object {
      "flex": 1,
      "padding": 10,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "space-between",
      }
    }
  >
    <View
      style={
        Object {
          "flex": 1,
          "flexDirection": "row",
          "justifyContent": "flex-start",
        }
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        disabled={false}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#000000",
            "fontFamily": "beleren",
          }
        }
      >
        CardName
      </Text>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "flex": 1,
          "flexDirection": "row",
          "justifyContent": "flex-end",
        }
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        disabled={false}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#000000",
            "fontFamily": "mtgx",
          }
        }
      >
        placeholdersToSymbols({1}{W}{B})
      </Text>
    </View>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "space-between",
      }
    }
  >
    <View
      style={
        Object {
          "flex": 1,
          "flexDirection": "row",
          "justifyContent": "flex-start",
        }
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        disabled={false}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#000000",
            "fontFamily": "mtgx",
          }
        }
      >
        CardType
      </Text>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "flex": 1,
          "flexDirection": "row",
          "justifyContent": "flex-end",
        }
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        disabled={false}
        ellipsizeMode="tail"
      >
        AST
      </Text>
    </View>
  </View>
  <View
    style={
      Object {
        "flexDirection": "row",
      }
    }
  >
    <View
      style={
        Object {
          "flex": 1,
          "flexDirection": "row",
        }
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        disabled={false}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#000000",
            "fontFamily": "mtgx",
          }
        }
      >
        CardText
      </Text>
    </View>
    <View
      style={
        Object {
          "alignItems": "flex-end",
          "flexDirection": "column-reverse",
        }
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        disabled={false}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#000000",
            "fontFamily": "mtgx",
          }
        }
      >
        2/4
      </Text>
    </View>
  </View>
</View>
`;
